### PR TITLE
cli: node groups

### PR DESCRIPTION
**Pull request by malt3**
### Context

This is the second part of nodeGroup enablement. While previously, we modified the terraform templates to support nodeGroups, this PR exposes nodeGroups to the user via the constellation-conf.yaml.

As a side-effect, we deprecate the "--control-plane-nodes" and "--worker-nodes" flags from "constellation create".

### Proposed change(s)

- add nodeGroups to config
- update config version to v4
- deprecate old flags
- remove deprecated flags from docs

### Future changes (for other PRs)

- [ ] add section in the docs to guide users on how to use node groups
- [ ] config migration (v3 -> v4)

### Checklist

- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [ ] Is PR title adequate for changelog?
- [x] Link to Milestone
- [ ] [e2e test](https://github.com/edgelesssys/constellation/actions/runs/5760853599)

### CODE REVIEW

File: internal/config/migration/migration.go
Comment (daniel-weisse) [inline internal/config/migration/migration.go:1]: The migrate command should migrate a config from the latest available release to a new format.
For example, v2.9.1 did not yet have the `CustomEndpoint` field in the config.
It might be worth to keep the migrate logic in a separate PR until we get closer to the next release and know what has to be included, and what can already be moved to the `v4` spec

Comment (malt3) [reply internal/config/migration/migration.go:1]: > The migrate command should migrate a config from the latest available release to a new format.
For example, v2.9.1 did not yet have the CustomEndpoint field in the config.

Agreed. I believe this is the only field and it is optional so this wouldn't be an issue but I'll remove it for clarity.

> It might be worth to keep the migrate logic in a separate PR until we get closer to the next release and know what has to be included, and what can already be moved to the v4 spec

I can do that. I think force pushing on this PR is required to make this less confusing.

File: dev-docs/howto/nfs.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ This document describes how to deploy NFS in Constellation using Rook.
 The cluster needs at least 3 worker nodes, default machines are powerful enough.
 
 ```bash
-constellation create --name nfs -c 1 -w 3
+constellation create --name nfs
```
Comment (daniel-weisse) [inline dev-docs/howto/nfs.md:10]:
```suggestion
constellation create
```
The name flag was also removed a while ago, no?

File: internal/config/config.go [hunk 2]
```diff
@@ -377,6 +369,24 @@ func Default() *Config {
...
 			QEMUVTPM:           &QEMUVTPM{Measurements: measurements.DefaultsFor(cloudprovider.QEMU, variant.QEMUVTPM{})},
 		},
+		NodeGroups: map[string]NodeGroup{
+			"control_plane_default": {
+				Role:            "control-plane",
+				Zone:            "",
+				InstanceType:    "",
+				StateDiskSizeGB: 30,
+				StateDiskType:   "",
+				InitialCount:    3,
```
Comment (elchead) [inline internal/config/config.go:379]: why 3 and only 1 worker? I think that 1 is a sensible default that a user would change if they need to?

**Review by daniel-weisse — APPROVED** [review]

**Comment by elchead** [comment]
Some doc comments:

- [ ]  how to migrate the config to `NodeGroups`.
- [ ] describe what values can be set for each field of the `NodeGroup`, specifically `Zone` (link to CSP), `InstanceType` and `StateDiskType`
- [ ] explicitly state that you can configure multiple node groups of the same role to allow different node setups
- [ ] state that at least a `control_plane_default` and `worker_default` nodeGroup are required (no renaming)

File: internal/config/config.go (hunk 2 above)
Comment (malt3) [reply internal/config/config.go:379]: 3 is better since it is the minimum number that ensures high availability of the control-plane (including etcd).
My reasoning is that we want to have defaults for a sane cluster that is unlikely to need recovery.
Users will see the amount of nodes clearly displayed when running "constellation create" so they can still change this to `1` if desired.

File: internal/config/validation.go [hunk 3]
```diff
@@ -129,6 +129,23 @@ func ValidateAWSZone(zone string) bool {
...
+
+// ValidateGCPZone validates that the zone is in the correct format.
+func ValidateGCPZone(zone string) bool {
+	gcpZoneRegex := regexp.MustCompile(`^[\w-]+-[a-z]$`)
+	return gcpZoneRegex.MatchString(zone)
+}
+
+// ValidateOpenStackRegion validates that the region is in the correct format.
+func ValidateOpenStackRegion(region string) bool {
+	return len(region) > 0
```
[the remark this answers is not in the document]
Comment (malt3) [reply internal/config/validation.go:146]: No. OpenStack is quite liberal in what a customer deployment use as a region naming scheme. [They even relax the allowed character set between versions](https://specs.openstack.org/openstack/nova-specs/specs/kilo/implemented/relax-resource-name-restrictions.html).

File: internal/config/config_test.go [hunk 4]
```diff
@@ -889,7 +944,7 @@ func TestValidateProvider(t *testing.T) {
 }
 
 func TestConfigVersionCompatibility(t *testing.T) {
-	t.Skip() // TODO(daniel-weisse): re-enable and re-write for config v3
+	t.Skip() // TODO(daniel-weisse): re-enable and re-write for config v4
```
[the remark this answers is not in the document]
Comment (malt3) [reply internal/config/config_test.go:947]: Done https://dev.azure.com/Edgeless/Edgeless/_workitems/edit/3326 :)

File: internal/config/migration/migration.go [hunk 5]
```diff
@@ -9,305 +9,12 @@ package migration
...
-	}
-
-	// Write migrated config
-	if err := fileHandler.WriteYAML(path, cfgV3, file.OptOverwrite); err != nil {
-		return fmt.Errorf("writing %s: %w", path, err)
-	}
-
-	return nil
+func V2ToV3(_ string, _ file.Handler) error {
+	// TODO(malt3): add migration from v3 to v4
```
Comment (derpsteb) [inline internal/config/migration/migration.go:18]: You are planning to add this in a separate PR?

Comment (malt3) [reply internal/config/migration/migration.go:18]: Yes. It was actually part of this PR and was moved out to make this PR more readable.

File: internal/config/config.go (hunk 2 above)
Comment (elchead) [reply internal/config/config.go:379]: Curious to learn why the number is 3 :)
Can we maybe keep the node flags for us devs internally (so that we can quickly set up a test cluster with 1 node to save costs)?
(I know we can workaround with yq but seems like an unnecessary step)

I definetly think that sensible defaults for the other fields should be set. So that `config generate` + `create` works out of the box

File: internal/config/config_test.go [hunk 6]
```diff
@@ -374,6 +393,24 @@ func TestValidate(t *testing.T) {
 				cnf.Attestation.AzureSEVSNP.Measurements = measurements.M{
 					0: measurements.WithAllBytes(0x00, measurements.Enforce, measurements.PCRMeasurementLength),
 				}
```
Comment (elchead) [inline internal/config/config_test.go:395]: nit: use modifyConfigForAzureToPassValidate here?

File: internal/config/config.go [hunk 7]
```diff
@@ -377,6 +369,24 @@ func Default() *Config {
 			GCPSEVES:           &GCPSEVES{Measurements: measurements.DefaultsFor(cloudprovider.GCP, variant.GCPSEVES{})},
 			QEMUVTPM:           &QEMUVTPM{Measurements: measurements.DefaultsFor(cloudprovider.QEMU, variant.QEMUVTPM{})},
 		},
+		NodeGroups: map[string]NodeGroup{
+			"control_plane_default": {
```
Comment (elchead) [inline internal/config/config.go:373]: nit
```suggestion
		constants.DefaultControlPlaneGroupName : {
```

File: internal/config/config.go [hunk 8]
```diff
@@ -377,6 +369,24 @@ func Default() *Config {
...
+				Role:            "control-plane",
+				Zone:            "",
+				InstanceType:    "",
+				StateDiskSizeGB: 30,
+				StateDiskType:   "",
+				InitialCount:    3,
+			},
+			"worker_default": {
```
Comment (elchead) [inline internal/config/config.go:381]:
```suggestion
			constants.DefaultWorkerGroupName: {
```

File: internal/config/config.go (hunk 2 above)
Comment (malt3) [reply internal/config/config.go:379]: Here are the details of why three is a good default: https://etcd.io/docs/v3.5/faq/#deployment also https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/#before-you-begin

> Can we maybe keep the node flags for us devs internally (so that we can quickly set up a test cluster with 1 node to save costs)?

I think as a developer you either have an existing config (with a customized number of initial nodes already set) or have to also set other config values (like debug, image). So I think we should not have a special flag and dogfood the sam UX that we want for our users. I also want to reduce the complexity of the cli.

File: docs/docs/workflows/create.md [hunk 9]
```diff
@@ -33,7 +33,7 @@ Choose the initial size of your cluster.
 The following command creates a cluster with one control-plane and two worker nodes:
```
Comment (elchead) [inline docs/docs/workflows/create.md:33]: remove

**Comment by malt3** [comment]
> For the regions / zones @thomasten said that we don't want to hardcode specific values in the validation, so that we can support new types without having to release a CLI. Is this different here?

@elchead what are you referring to? I think all of the regexes used for region / zone validation are already designed to allow new zones (as long as they follow the naming scheme used by the CSPs).
In case you are referring to the disk types: Those were already hardcoded in the config before. I just moved this.
I think there is no clear pattern for the disk type names so we should either keep the strict validation or remove it entirely.

**Comment by elchead** [comment]
> > For the regions / zones @thomasten said that we don't want to hardcode specific values in the validation, so that we can support new types without having to release a CLI. Is this different here?
> 
> @elchead what are you referring to? I think all of the regexes used for region / zone validation are already designed to allow new zones (as long as they follow the naming scheme used by the CSPs). In case you are referring to the disk types: Those were already hardcoded in the config before. I just moved this. I think there is no clear pattern for the disk type names so we should either keep the strict validation or remove it entirely.

Yeah referring to disk + instance types. I prefer helping the user with the options but would just want @thomasten to confirm that this is what we want

File: cli/internal/terraform/variables.go [hunk 10]
```diff
@@ -74,8 +74,8 @@ type AWSNodeGroup struct {
 	// StateDiskSizeGB is the size of the state disk to allocate to each node, in GB.
 	StateDiskSizeGB int `hcl:"disk_size" cty:"disk_size"`
 	// InitialCount is the initial number of nodes to create in the node group.
-	// During upgrades this value is not set.
-	InitialCount *int `hcl:"initial_count" cty:"initial_count"`
+	// During upgrades this value ignored.
+	InitialCount int `hcl:"initial_count" cty:"initial_count"`
```
[the remark this answers is not in the document]
Comment (malt3) [reply cli/internal/terraform/variables.go:78]: I found it rather confusing to use since the same variable is passed around over multiple layers and is only considered a pointer type here.
The name makes it very obvious that this is only used for the *initial* count so I believe this is easier and safer (cannot be nil, no chance to deref a nil ptr).

File: internal/config/validation.go [hunk 11]
```diff
@@ -254,8 +367,53 @@ func (c *Config) translateMoreThanOneAttestationError(ut ut.Translator, fe valid
 	return t
 }
 
+func registerTranslateDiskTypeError(ut ut.Translator) error {
```
[the remark this answers is not in the document]
Comment (malt3) [reply internal/config/validation.go:370]: I think this is fine because the error message already gives you a list of all allowed values.
Before, the URLs where part of the config yaml comments and we did not have an exhaustive list of allowed values in that comment.

**Comment by malt3** [comment]
> Some doc comments:
> 
> * [ ]   how to migrate the config to `NodeGroups`.
> * [ ]  describe what values can be set for each field of the `NodeGroup`, specifically `Zone` (link to CSP), `InstanceType` and `StateDiskType`.
> * [ ]  explicitly state that you can configure multiple node groups of the same role to allow different node setups
> * [ ]  ~state that at least a `control_plane_default` and `worker_default` nodeGroup are required (no renaming)~ since we validate the config, it seems easier to maintain only the code

@elchead this will be addressed in a separate PR as pointed out in the PR description:

> add section in the docs to guide users on how to use node groups


I now also created AB#3326 to track this.

File: docs/docs/getting-started/first-steps.md [hunk 12]
```diff
@@ -128,7 +128,7 @@ If you encounter any problem with the following steps, make sure to use the [lat
     :::
```
Comment (thomasten) [inline docs/docs/getting-started/first-steps.md:128]: fix "Create the cluster with one control-plane node and two worker nodes" above

File: docs/docs/getting-started/first-steps.md [hunk 13]
```diff
@@ -128,7 +128,7 @@ If you encounter any problem with the following steps, make sure to use the [lat
...
 
     ```bash
-    constellation create --control-plane-nodes 1 --worker-nodes 2 -y
+    constellation create -y
     ```
 
     This should give the following output:
```
Comment (thomasten) [inline docs/docs/getting-started/first-steps.md:134]: replace `$ constellation create ...` with `$ constellation create -y` below

File: docs/docs/getting-started/first-steps-local.md [hunk 14]
```diff
@@ -103,7 +103,7 @@ attaching persistent storage, or autoscaling aren't available.
 2. Now you can create your cluster and its nodes. `constellation create` uses the options set in `constellation-conf.yaml`.
 
   ```bash
-  constellation create --control-plane-nodes 1 --worker-nodes 1
+  constellation create
   ```
 
   This will create 2 VMs: one worker node, and one control plane node.
```
Comment (thomasten) [inline docs/docs/getting-started/first-steps-local.md:109]: fix

File: docs/docs/workflows/create.md [hunk 15]
```diff
@@ -29,11 +29,10 @@ Before you create the cluster, make sure to have a [valid configuration file](./
 <tabs groupId="provider">
 <tabItem value="cli" label="CLI">
 
-Choose the initial size of your cluster.
-The following command creates a cluster with one control-plane and two worker nodes:
+Choose the initial size of your cluster by setting the `initialCount` field in the `constellation-config.yaml`.
```
Comment (thomasten) [inline docs/docs/workflows/create.md:32]: I think this doesn't belong here, but in the Configure workflow

File: docs/docs/workflows/create.md [hunk 16]
```diff
@@ -29,11 +29,10 @@ Before you create the cluster, make sure to have a [valid configuration file](./
...
 
 ```bash
-constellation create --control-plane-nodes 1 --worker-nodes 2
+constellation create
 ```
 
 For details on the flags, consult the command help via `constellation create -h`.
```
Comment (thomasten) [inline docs/docs/workflows/create.md:38]:
```suggestion
```
remaining flags are not that interesting

File: docs/docs/workflows/troubleshooting.md [hunk 17]
```diff
@@ -31,7 +31,7 @@ Afterward, set `ARM_SKIP_PROVIDER_REGISTRATION=true` as an environment variable
 For example:
 
 ```bash
-ARM_SKIP_PROVIDER_REGISTRATION=true constellation create --control-plane-nodes 1 --worker-nodes 2 -y
+ARM_SKIP_PROVIDER_REGISTRATION=true constellation create -y
```
Comment (thomasten) [inline docs/docs/workflows/troubleshooting.md:34]:
```suggestion
ARM_SKIP_PROVIDER_REGISTRATION=true constellation create
```
nit: I'd prefer to omit the flag for this example

**Review by thomasten — CHANGES_REQUESTED** [review]

**Review by derpsteb — APPROVED** [review]
Can't remove myself from PR :(
Didn't look in depth.

**Review by thomasten — APPROVED** [review]
